include/x11/background_manager.hpp

```diff
@@ -42,48 +83,35 @@ class background_manager : public signal_receiver<SIGN_PRIORITY_SCREEN, signals:
    * Starts observing a rectangular slice of the desktop background.
    *
    * After calling this function, you can obtain the current slice of the desktop background
-   * with background_manager::get_surface. Whenever the background slice changes (for example,
-   * due to bar position changes or because the user changed the desktop background) the class
-   * emits a signals::ui::update_background event.
+   * by calling get_surface on the returned bg_slice object.
+   * Whenever the background slice changes (for example, due to bar position changes or because
+   * the user changed the desktop background) the class emits a signals::ui::update_background event.
    *
-   * \param window This should be set to the bar window.
-   * \param rect Slice of the background to observe (coordinates relative to window).
-   *             Typically set to the outer area of the bar.
-   */
-  void activate(xcb_window_t window, xcb_rectangle_t rect);
-
-
-  /**
-   * Stops observing the desktop background and frees all resources.
-   */
-  void deactivate();
-
-  /**
-   * Retrieve the current desktop background slice.
+   * You should only call this function once and then re-use the returned bg_slice because the bg_slice
+   * caches the background. If you don't need the background anymore, destroy the shared_ptr to free up
+   * resources.
    *
-   * This function returns a slice of the desktop background that has the size of the rectangle
-   * given to background_manager::activate. As the slice is cached by the manager, this function
-   * is fast.
+   * \param rect Slice of the background to observe (coordinates relative to window).
+   * \param window Coordinates are interpreted relative to this window
    */
-  cairo::surface* get_surface() const;
+  std::shared_ptr<bg_slice> observe(xcb_rectangle_t rect, xcb_window_t window);
 
   void handle(const evt::property_notify& evt);
   bool on(const signals::ui::update_geometry&);
  private:
+  void activate();
+  void deactivate();
+
   // references to standard components
   connection& m_connection;
   signal_emitter& m_sig;
   const logger& m_log;
 
-  // these are set by activate
-  xcb_window_t m_window;
-  xcb_rectangle_t m_rect{0, 0, 0U, 0U};
+  // list of slices that need to be filled with the desktop background
+  std::vector<std::weak_ptr<bg_slice>> m_slices;
```

Comment:
Why do we use `weak_ptr` here? Would `shared_ptr` not also work?

Reply:
the idea is that when a component does not require it's slice anymore, we will remove it from the list. To detect this, we need a weak_ptr (since we don't want to keep the slice alive if no component uses it anymore).

Reply:
Ah, is that also the reason you didn't use `make_shared` to create new slices?

Reply:
the reason for not using `make_shared` is that the constructor is private (only the manager should be able to create slices) and `make_shared` would require it to be public.

Reply:
Didn't even notice that it was private